### ISSUE TITLE
Possible silent stalling

### ISSUE THREAD

**Issue by michaeljmetzger** [KamilSJaron/smudgeplot#40]
I am trying to run smudgeplot hetkmers on a server with the following command: 

```
smudgeplot hetkmers -o MELC-2E11_R1_trim30_KMC21_10-950_pairs < MELC-2E11_R1_trim30_KMC21_10-950.txt
```
The output I have so far is:

```
Running smudgeplot v0.2.0
Task: hetkmers
Kmers and coverages loaded.
```

This showed up relatively quickly upon executing the command, but that was four days ago. There have been no new files generated in the working directory, and there has been no output to the terminal suggesting any continued operation. Is it likely there has been a silent error? Would you expect any output showing continued progress or is it expected to be silent until completion? If it is likely to be an error, do you have any thought on what that could be?

Thanks,
Michael

**Comment by KamilSJaron** [comment]
Hello Michael,

having some sort of a progress bar would be certainly nice and perhaps it won't be even too complicated to implement.

Right now we are not aware of any bug in the code, therefore my first guess would be that your file is just really big. How many kmers you have in the dump file? (something like `wc -l MELC-2E11_R1_trim30_KMC21_10-950.txt` if really big, you can just share the size of the file).

**Comment by KamilSJaron** [comment]
Ha, one more thought. What `L` you have chosen to generate the dump file? Could you post here also the kmer distribution? (i.e. genomescope)

**Comment by michaeljmetzger** [comment]
Thanks so much for your response. Even if there is no progress bar, a timestamp with a "still working" output every so often might be useful for cases like this.

The L and U are 10 and 950, and there are 900591696 kmers in the file, using the wc code you posted above. 

I expect this sample to be a normal diploid clam genome. I do expect a lot of heterozygosity, even for this normal sample. The genomescope profile (run on the full 1-10000 coverage dump file) gives 1.79% heterozygosity. My next samples to run will be cancer genomes, which I expect to be super weird (and roughly tetraploid), so --middle might not work well for me in the long run.

If you think silence means something is happening then I guess I can just let it run a bit longer. It does make me nervous that no files have been created in nearly a week of running, so I am still suspicious that it has stalled somewhere. If you can think of anything that might help, please let me know. I was hoping this would give me some interesting results prior to assembly of the full reference genome, but smudgeplot may be more complicated and computationally intensive than genome assembly itself, so I may have to abandon it soon so I can use our server for something that looks like it has a better chance of success. 

![MELC-2E11-trim30-k21-GSplot](https://user-images.githubusercontent.com/51342699/60612524-a9fe5100-9d7d-11e9-8735-a587e8bb6a71.png)

Thanks for your advice!
--Michael

**Comment by KamilSJaron** [comment]
This kmer profile looks beautiful. The sequencing is not very deep, but the peaks are very nicely separated. The only thing that smudgeplot could possibly reveal is how many (homologous) kmer pairs are within the diploid peak (i.e. AABB smudge), which I suspect to be very low compared to the heterozygous smudge (AB) formed from kmers found in the heterozygosity peak (21x). I am huge fan of the approach you have chosen - to take a look before assembling it. However in this case I don't expect smudgeplot to bring much insights. It's more build for the weird critter (I guess your cancers are actually the arhropods).

900k of kmers is really a lot. I am not that surprised that it takes a while. Also L = 10 is rather low (I expect that cutoff algorithm has suggested it). However, looking at the kmer spectra, I don't think it will hange much (you should always try to minimize numer of error kmers in the dump file).

Finally, the reason why the algorithm is not generating results on the fly is that we first build up the groups of kmers and then we keep only those that form a unique pair. Which means that for this filtering we need to sort of out all the kmers first. At the same time you are right, there is no reason why not to print a progress bar.

**Comment by michaeljmetzger** [comment]
Thanks for the quick response! This is the normal genome, so it is really the control. I expect the polyploid cancer genome to be much weirder--that is what I really want to see on smudgeplot. If there is reason to believe it is still running that is ok--I just don't want to keep staring at a blank terminal for a month only to find out there was a silent error on day 1.

Any suggestions for reducing kmer count? Would you recommend redoing at a lower kmer size (like 17 instead of 21) or making more strict cutoffs (like 12 & 600) or do you think I would lose too much data doing that?

**Comment by michaeljmetzger** [comment]
Quick followup question: Would I be able to use the -t option to increase the number of threads used? That is used for KMC but I didn't see an analogous option for smudgeplot hetkmers. If I am reading the processor readings correctly, python is only using one thread.

**Comment by KamilSJaron** [comment]
re silent error: No silent error was reported until now and the script is rather simple. Honestly, I don't think it's a silent error.

re reducing number of kmers: lowering k helps, increasing L as well. I would not be worried to even use something like 15. You will filter out plenty of genomic kmers, but there are soooo many that I would be alright. Basically with L you always want to be bellow expected 1n coverage. Are this all the data you have by the way?

re multithreading: We implemented multithreading before, but there were a few issues. Somehow the cores were not used all the time but more important, in big datasets we had troubles to close computing threads (#18 was due to multithreading). Therefore we decided to turn hetkmers singlethreded again #27. (if there would be someone who would like to try to make it parallel we are happy to accept pull requests)

**Comment by michaeljmetzger** [comment]
I may retry with a lower kmer size since that should drastically reduce the number. I believe that kmer file was from only the first read of a paired end sequencing run, so I can increase a bit by concatenation those two reads. I have data from several more individuals, but they are all at roughly the same coverage.

Thanks again for your advice.

Best,
Michael

**Comment by KamilSJaron** [comment]
Happy to help.

Use only a single individual, but all the data you have (more details are [here](https://github.com/KamilSJaron/smudgeplot/wiki/everything-about-input-data)). You don't even have to catinate it, you can just tell KMC to calculate the database from multiple files. Finally, if you have done trimming already, you can run it on the trimmed dataset as well (discussed in #3 ). If you would not be sure about choosing L, you can post here the histogram you get from all the data.

**Comment by michaeljmetzger** [comment]
Thanks again for your help. How would you like usage of smudgeplot to be cited if used in a in a seminar (On the off chance this hetkmers run finishes by Friday)?
--Michael

**Comment by KamilSJaron** [comment]
Wait, hetkmers is still running? That's frustrating! Perhaps there is something in your data that gives the algorithm hard times or even worse, maybe you were right in the first place (that the algorithm is silently stalling). Does the program run for more than a week? If yes, would you be willing to either share the data or somehow help me understand what is going on?

re citaiton: Right now people just give the url of this repo. Unfortunately, we don't have the preprint out yet. However, thanks for acknowledging us.

**Comment by michaeljmetzger** [comment]
I redid the analysis with k17 instead of k21 and much more strict L and U values, and this one has only been running for 3-4 days now so far.

I will make sure to acknowledge you if I talk about the data.

Thanks,
--Michael

**Comment by michaeljmetzger** [comment]
Smudgeplot has now been running with the much smaller file for more than two weeks with no progress beyond :Task: hetkmers.” I now no longer see python listed as a running process when using the “top” command. I would be happy to share data or try any other suggestions.
--Thanks,
Michael



From: Michael Metzger
Sent: Wednesday, July 17, 2019 6:33 PM
To: KamilSJaron/smudgeplot <reply@reply.github.com>; KamilSJaron/smudgeplot <smudgeplot@noreply.github.com>
Cc: Author <author@noreply.github.com>
Subject: RE: [KamilSJaron/smudgeplot] Possible silent stalling (#40)

I redid the analysis with k17 instead of k21 and much more strict L and U values, and this one has only been running for 3-4 days now so far.

I will make sure to acknowledge you if I talk about the data.

Thanks,
--Michael

**Comment by KamilSJaron** [comment]
Dang!

Then it must be a problem in the program. Could you share the small dump file please? I will find out where is the problem...

Thank you very much for your patience.

**Comment by michaeljmetzger** [comment]
I am happy to share the dump file if you would be able to take a look at what could be going wrong. Do you have a preferred way for me to send that file? It is 12 GB.

--Michael

**Comment by KamilSJaron** [comment]
Thanks a lot, but I am afraid we don't have a good standartized way.

If you have an ftp server where you can upload it, that would be perfect. Otherwise, could google drive or dropbox work? Or some of the file sharing services?

I would also suggest to compress the file. It's a plain text and it has a really good compresibility.

**Comment by KamilSJaron** [comment]
-> progress bar would be useful, but impractical to implement as the algorithm is recursive. (adding won't fix)

-> I try to reproduced the stalling behaviour with the shared files (thanks for that)

**Comment by KamilSJaron** [comment]
I think I can with certainty say it's a lot more likely I will write a completely new algorithm to find new pairs than return to this error.

I actually reproduced the behavior if I remember right (that it took a lot longer than expected), but I never figured out what is going on. Really sorry I did not get around to fix it, but I feel like at this moment (that we have the kmc version of finding kmer pairs) there is not much point in digging that much in it.
